### ISSUE TITLE
How to check if a generic type is an instance of another generic type?

### ISSUE THREAD

**Issue by TypeDefinition** [nus-cs2030/2021-s1#70]
# Description 
In Lab 4 Part 5 we are tasked with throwing an ArrayStoreException when an array of the wrong type is being passed as the argument. One way I can think of is to check if type T is an instance of the argument type. However, due to type erasure, there is no way to check for it.

I know I can complete this question by simply wrapping it in a try catch, catching the error Java throws and rethrowing it with our own message. But I would like to know how Java itself manages to detect the error and throw the exception.

## Screenshots (if any): 
![image](https://user-images.githubusercontent.com/13763067/93694052-df858a80-fb39-11ea-8e0e-11a6f5d1b6ef.png)

**Comment by domlimm** [comment]
Hi @TypeDefinition ! Not sure if #66 would help. Could be similar to what you're looking for.

**Comment by TypeDefinition** [comment]
Hi @domsterthebot, thanks bro but as stated in #66 itself, I don't think it works for Lab4. I've tried it with normal types, which works, but when it comes to generics, it no longer works. The reason I think is due to type erasure. From what I understand, basically everything that is a generic is replaced with an Object reference, and casted to whatever type. This means that Type T is actually an Object array, and when we pass in an Integer Array, an Integer is NOT a super of Object.

**Comment by TypeDefinition** [comment]
Also, any implementation that requires getting an element of the list is unsafe as there is no guarantee that the lists are not empty.

**Comment by SkullPirateTK** [comment]
I looked into the jdk11 source code for a bit, it seems that the exception is thrown from the native interface `java.lang.System.arraycopy()`(used by `java.util.Arrays.copyOf()`) which is from another language (probably c/c++). So I suppose that there is no way to catch the error in Java itself.

**Comment by TypeDefinition** [comment]
Alrighty, I found the answer.

TLDR: Even the creators of Java cannot tell you if the array can store the elements if the ArrayList is empty.
So if they can't do it, I'm gonna assume there is just no way.

Test Case:
![image](https://user-images.githubusercontent.com/13763067/93708468-04452680-fb69-11ea-903f-556767e7e5c1.png)

Code from OpenJDK's ArrayList.toArray:
![image](https://user-images.githubusercontent.com/13763067/93708535-964d2f00-fb69-11ea-8288-fa7e52a43092.png)

Code from OpenJDK's Arrays.copyOf:
![image](https://user-images.githubusercontent.com/13763067/93708556-c5fc3700-fb69-11ea-9dc9-de5ea3aa32d6.png)

Code from OpenJDK's System.arraycopy (Code isn't implemented in Java):
![image](https://user-images.githubusercontent.com/13763067/93708910-a1558e80-fb6c-11ea-872f-e0bda48203c3.png)

**Comment by domlimm** [comment]
> Hi @domsterthebot, thanks bro but as stated in #66 itself, I don't think it works for Lab4. I've tried it with normal types, which works, but when it comes to generics, it no longer works. The reason I think is due to type erasure. From what I understand, basically everything that is a generic is replaced with an Object reference, and casted to whatever type. This means that Type T is actually an Object array, and when we pass in an Integer Array, an Integer is NOT a super of Object.

Hi @TypeDefinition , sorry about that...! I've been stuck on this too till 4 last night. Still haven't found a solution... I was just looking at your latest comment. Any pointers on this? 😢 Think we need a workaround

**Comment by SkullPirateTK** [comment]
> > Hi @domsterthebot, thanks bro but as stated in #66 itself, I don't think it works for Lab4. I've tried it with normal types, which works, but when it comes to generics, it no longer works. The reason I think is due to type erasure. From what I understand, basically everything that is a generic is replaced with an Object reference, and casted to whatever type. This means that Type T is actually an Object array, and when we pass in an Integer Array, an Integer is NOT a super of Object.
> 
> Hi @TypeDefinition , sorry about that...! I've been stuck on this too till 4 last night. Still haven't found a solution... I was just looking at your latest comment. Any pointers on this? 😢 Think we need a workaround

The way to solve it is actually quite simple, as Java will do dynamic typing during runtime. The overloaded toArray() basically just takes an Object[] and returns an Object[] using List.toArray(T[] arr).

**Comment by domlimm** [comment]
> > > Hi @domsterthebot, thanks bro but as stated in #66 itself, I don't think it works for Lab4. I've tried it with normal types, which works, but when it comes to generics, it no longer works. The reason I think is due to type erasure. From what I understand, basically everything that is a generic is replaced with an Object reference, and casted to whatever type. This means that Type T is actually an Object array, and when we pass in an Integer Array, an Integer is NOT a super of Object.
> > 
> > 
> > Hi @TypeDefinition , sorry about that...! I've been stuck on this too till 4 last night. Still haven't found a solution... I was just looking at your latest comment. Any pointers on this? 😢 Think we need a workaround
> 
> The way to solve it is actually quite simple, as Java will do dynamic typing during runtime. The overloaded toArray() basically just takes an Object[] and returns an Object[] using List.toArray(T[] arr).

But how will we be better to check on the `ArrayStoreException` error?

I keep facing this issue,
![image](https://user-images.githubusercontent.com/46486515/93736147-fa2b3280-fc11-11ea-8ae1-fa71ffe1d62f.png)

Like you mentioned, previously i was using generic type, T, for my input param and the return type as `T[...]`

Changed to:

```java
    public Object[] toArray(Object[] data) {
        if (data == null) {
            throw new NullPointerException("Input array cannot be null");
        }
        
        List<Object> newList = new ArrayList<Object>(this.list);

        if (newList.toArray(data) instanceof Object[] != true) {
            throw new ArrayStoreException("Cannot add element to array as it is the wrong type");
        }

        return newList.toArray(data);
    }
```

Now, i got this,
![image](https://user-images.githubusercontent.com/46486515/93736542-432fb680-fc13-11ea-9710-69b6996219f5.png)

**Comment by SkullPirateTK** [comment]
Wrap it in a try catch block instead using if statements to preempt the error.

**Comment by domlimm** [comment]
Ah damn... I tried it after i commented and probably wrote it wrongly. Thank you so much! 😃  @SkullPirateTK

EDIT:
My final overloaded method `toArray(Object[] o)`
```java
public Object[] toArray(Object[] data) {
    try {
        return this.list.toArray(data);
    }
    catch (ArrayStoreException e) {
        throw new ArrayStoreException("Cannot add element to array as it is the wrong type");
    }
    catch (NullPointerException e) {
        throw new NullPointerException("Input array cannot be null");
    }
}
```

**Comment by TypeDefinition** [comment]
@domsterthebot

Basically the only way we can do it is to call the built in ArrayList.toArray method. We will then wrap it around a try catch. When ArrayList.toArray throws an ArrayStoreException, we will catch it, and throw our own ArrayStoreException with the overidden message. As for detecting the problem on our end, I do not believe there is any way to do it in Java. Even Java's built in ArrayList.toArray, requires you to have at least 1 element in the ArrayList for it to detect the class type.

If you see the above code snippets, if you give an empty ArrayList<Integer> a String array as input, it will not throw you ArrayStoreException, since it probably cannot detect that the String array is invalid. 

Java's actual ArrayList:
![Untitled](https://user-images.githubusercontent.com/13763067/93738127-90ae2280-fc17-11ea-8b4d-c9b648952f82.png)

This is the best emulation I can come up with. This works basically like ArrayList.toArray:
![image](https://user-images.githubusercontent.com/13763067/93737998-3e6d0180-fc17-11ea-8a01-9bec6faf8fad.png)

**Comment by TypeDefinition** [comment]
@domsterthebot @SkullPirateTK 

I have found where the exception is thrown.
[](https://stackoverflow.com/questions/12594046/java-native-method-source-code)
It was implemented in C++.

**Comment by TypeDefinition** [comment]
@SkullPirateTK @domsterthebot 

I believe the exception is actually thrown by the native System.arraycopy function which is implemented in C++.

https://github.com/openjdk/jdk/blob/master/src/hotspot/share/oops/objArrayKlass.cpp (Line 238)

**Comment by domlimm** [comment]
> @SkullPirateTK @domsterthebot
> 
> I believe the exception is actually thrown by the native System.arraycopy function which is implemented in C++.
> 
> https://github.com/openjdk/jdk/blob/master/src/hotspot/share/oops/objArrayKlass.cpp (Line 238)

Sorry didn't want to flood this thread by quoting your answer @TypeDefinition . Yes i get what you mean! I did research too and there's no way to fix this. Good one with what you've done by using this method (your `toArray(U[] _input)` of greater length) to specify another generic name/type (what do they call it? haha) and then iterate through `this.list` and compare the data type for each element. +1 for that!

I kind of dropped `getClass()` after seeing that it causes Java reflection. In #66 someone mentioned to use `isAssignableFrom()`. Good use of that, nice catch!!

> > > Hi @domsterthebot, thanks bro but as stated in #66 itself, I don't think it works for Lab4. I've tried it with normal types, which works, but when it comes to generics, it no longer works. The reason I think is due to type erasure. From what I understand, basically everything that is a generic is replaced with an Object reference, and casted to whatever type. This means that Type T is actually an Object array, and when we pass in an Integer Array, an Integer is NOT a super of Object.
> > 
> > 
> > Hi @TypeDefinition , sorry about that...! I've been stuck on this too till 4 last night. Still haven't found a solution... I was just looking at your latest comment. Any pointers on this? 😢 Think we need a workaround
> 
> The way to solve it is actually quite simple, as Java will do dynamic typing during runtime. The overloaded toArray() basically just takes an Object[] and returns an Object[] using List.toArray(T[] arr).

Maybe @SkullPirateTK can shed some light and elaborate further on how this below works! I can't really wrap my head around this method above.

Thank you to you both! @TypeDefinition @SkullPirateTK

**Comment by SkullPirateTK** [comment]
@domsterthebot
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/List.html#toArray(T%5B%5D)
When you pass an array of any type into the function `Object[] toArray(Object[] data)`, it is legal because `Object[]` is the super class any array (that is not using a primitive type). Then after passing `data` into `List.toArray()`, `List.toArray(data)` returns an array of the same type as `data`, which is then returned as the output of `Object[] toArray(Object[] data)`. Since the array from `toArray(data)` will always be a subclass of `Object[]`, it will always work.

**Comment by domlimm** [comment]
> @domsterthebot
> https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/List.html#toArray(T%5B%5D)
> When you pass an array of any type into the function `Object[] toArray(Object[] data)`, it is legal because `Object[]` is the super class any array (that is not using a primitive type). Then after passing `data` into `List.toArray()`, `List.toArray(data)` returns an array of the same type as `data`, which is then returned as the output of `Object[] toArray(Object[] data)`. Since the array from `toArray(data)` will always be a subclass of `Object[]`, it will always work.

Ah.. Crystal clear! Then the `try { ... } catch { ... }` will supposedly handle any issues. Sweet! Thanks so much!!! @SkullPirateTK

**Comment by aaron-ljx** [comment]
> 
> 
> @domsterthebot
> https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/List.html#toArray(T%5B%5D)
> When you pass an array of any type into the function `Object[] toArray(Object[] data)`, it is legal because `Object[]` is the super class any array (that is not using a primitive type). Then after passing `data` into `List.toArray()`, `List.toArray(data)` returns an array of the same type as `data`, which is then returned as the output of `Object[] toArray(Object[] data)`. Since the array from `toArray(data)` will always be a subclass of `Object[]`, it will always work.

An alternate way of looking at it is that arrays are covariant in Java, meaning you can do something like this.

![WhatsApp Image 2020-09-22 at 02 56 52](https://user-images.githubusercontent.com/70706659/93809039-5c6e4c80-fc7f-11ea-887e-ee6dd78e1e37.jpeg)

Since arrays are covariant as mentioned above, the array with type U can be upcasted as an array with type Object since U is a subclass of Object. Since our type U array has been upcasted, the compiler treats that array as if it is an type Object array and so the compiler won't complain even if you put an element with type T into that array. However, arrays are immune to type erasure at runtime and so it will remember that it is actually a type U array and throws an ArrayStoreException if you put in an element of a different type.

Here are some links below if you're interested in more details.

https://stackoverflow.com/questions/24232141/why-is-upcasting-arrays-allowed-while-not-with-generics
https://stackoverflow.com/questions/18666710/why-are-arrays-covariant-but-generics-are-invariant
https://stackoverflow.com/questions/23414090/what-is-the-difference-between-up-casting-and-down-casting-with-respect-to-class

**Comment by domlimm** [comment]
Thanks @aaron-ljx !
